unstructured_ingest/v2/processes/connectors/vectara.py

```diff
@@ -0,0 +1,310 @@
+    token_url: str = "https://vectara-prod-{}.auth.us-west-2.amazoncognito.com/oauth2/token"
+
+
+class VectaraUploadStagerConfig(UploadStagerConfig):
+    pass
+
+
+@dataclass
+class VectaraUploadStager(UploadStager):
+    upload_stager_config: VectaraUploadStagerConfig = field(
+        default_factory=lambda: VectaraUploadStagerConfig()
+    )
+
+    @staticmethod
+    def conform_dict(data: dict) -> dict:
+        """
+        Prepares dictionary in the format that Vectara requires.
+        See more detail in https://docs.vectara.com/docs/rest-api/create-corpus-document
+
+        Select which meta-data fields to include and optionally map them to a new format.
+        remove the "metadata-" prefix from the keys
+        """
+        metadata_map = {
+            "page_number": "page_number",
+            "data_source-url": "url",
+            "filename": "filename",
+            "filetype": "filetype",
+            "last_modified": "last_modified",
+            "element_id": "element_id",
+        }
+        md = flatten_dict(data, separator="-", flatten_lists=True)
+        md = {k.replace("metadata-", ""): v for k, v in md.items()}
+        md = {metadata_map[k]: v for k, v in md.items() if k in metadata_map}
+        return md
+
+    def run(
+        self,
+        elements_filepath: Path,
+        file_data: FileData,
+        output_dir: Path,
+        output_filename: str,
+        **kwargs: Any,
+    ) -> Path:
+        with open(elements_filepath) as elements_file:
+            elements_contents = json.load(elements_file)
+
+        logger.info(
+            f"Extending {len(elements_contents)} json elements from content in {elements_filepath}"
+        )
+
+        docs_list = []
+
+        conformed_elements = {
+            "id": str(uuid.uuid4()),
+            "type": "core",
+            "metadata": {
+                "title": file_data.identifier,
+            },
+            "document_parts": [
+                {
+                    "text": element.pop("text", None),
+                    "metadata": self.conform_dict(data=element),
+                }
+                for element in elements_contents
+            ],
+        }
+
+        docs_list.append(conformed_elements)
+
+        output_path = Path(output_dir) / Path(f"{output_filename}.json")
+        with open(output_path, "w") as output_file:
+            json.dump(docs_list, output_file)
+        return output_path
+
+
+class VectaraUploaderConfig(UploaderConfig):
+    pass
+
+
+@dataclass
+class VectaraUploader(Uploader):
+
+    connector_type: str = CONNECTOR_TYPE
+    upload_config: VectaraUploaderConfig
+    connection_config: VectaraConnectionConfig
+    _jwt_token: Optional[str] = field(init=False, default=None)
+    _jwt_token_expires_ts: Optional[float] = field(init=False, default=None)
+
+    def is_async(self) -> bool:
+        return True
+
+    def precheck(self) -> None:
+        try:
+            asyncio.run(self._check_connection_and_corpora())
```

Comment:
This actually breaks when the precheck is run within an async context, because `asyncio.run` tried to create a new one. Ideally if we can run this check without any async code, that would be ideal. Such as supporting an async and non-async client.